File: Graph/csharp/ConnectedComponentsCount.cs
using System.Collections.Generic;

namespace GraphSolutions;

public static class ConnectedComponentsCount
{
    public static int Solve(Dictionary<string, List<string>> graph)
    {
        var visited = new HashSet<string>();
        var count = 0;

        foreach (var node in graph.Keys)
        {
            if (Explore(graph, node, visited))
            {
                count++;
            }
        }

        return count;
    }

    private static bool Explore(Dictionary<string, List<string>> graph, string node, HashSet<string> visited)
    {
        if (!visited.Add(node))
        {
            return false;
        }

        if (!graph.TryGetValue(node, out var neighbors))
        {
            return true;
        }

        foreach (var neighbor in neighbors)
        {
            Explore(graph, neighbor, visited);
        }

        return true;
    }
}

File: Graph/php/connectedComponentsCount.php
<?php

declare(strict_types=1);

function connectedComponentsCount(array $graph): int
{
    $visited = [];
    $count = 0;

    foreach ($graph as $node => $_) {
        if (exploreComponent($graph, (string) $node, $visited)) {
            $count++;
        }
    }

    return $count;
}

function exploreComponent(array $graph, string $node, array &$visited): bool
{
    if (isset($visited[$node])) {
        return false;
    }
    $visited[$node] = true;

    foreach ($graph[$node] ?? [] as $neighbor) {
        exploreComponent($graph, (string) $neighbor, $visited);
    }

    return true;
}
